lib/python/picongpu/extra/utils/picmi_density_reader.py

```diff
@@ -0,0 +1,144 @@
+        index_first_c = s.find(":")
+        if index_first_q == -1 and index_first_c == -1:
+            return False
+        else:
+            return True
+
+    def inner_evaluate(self, s, x, symbol):
+        """evalute string - this is a recursevly called internal method
+        s ... string code
+        x ... float value to evalute density at
+        symbol ... string symbol to replace
+        """
+        if self.test_for_cases(s):
+            if self.debug:
+                print("CASES")
+            return self.eval_cases(s, x, symbol)
+
+        else:
+            s_sym = sympy.parsing.sympy_parser.parse_expr(s)
+            res = s_sym.subs(symbol, x)
+            if self.debug:
+                print("eval:", res)
+            return res
+
+    def eval_cases(self, s, x, symbol):
+        """handle c++ cases of form condition ? case1 : case 2
+        s ... string code
+        x ... float value to evalute
+        symbol ... symbol to replace in string s
+        """
+        if self.debug:
+            print("-->", s)
+        s = self.clean_substring(s)
+        if self.debug:
+            print("==>", s)
+        index_first_q = s.find("?")
+        index_first_c = s.find(":")  # this assumes that there are no cases in the first branch
+
+        condition = s[0:index_first_q]
+        case1 = s[index_first_q + 1 : index_first_c]
+        case2 = s[index_first_c + 1 :]
+
+        if self.debug:
+            print("if")
+            print(condition)
+            print("do")
+            print(case1)
+            print("else")
+            print(case2)
+            print("fi")
+
+        if self.inner_evaluate(condition, x, symbol):
+            return self.inner_evaluate(case1, x, symbol)
+        else:
+            return self.inner_evaluate(case2, x, symbol)
+
+    def initial_clean(self, s):
+        """initially clean string from c++/picongpu methods
+        making it readably for sympy
+        s ... string
+        """
+        if self.debug:
+            print("clean before:", s)
+        s = s.replace("^", "**")
+        s = s.replace("pmacc::math::", " ")
+        if self.debug:
+            print("clean after:", s)
+        return s
+
+    def clean_substring(self, s):
+        """clean any substrings from cases from paraneties and whitespace
+        s ... string
+        """
+        s.strip()
+
+        while s.startswith("("):
+            if not s.endswith(")"):
+                raise ValueError(f"Missing closing parenthesis in {s=}.")
+            s = s[1:-1]
+        return s
+
+
+if __name__ == "__main__":
+    # load pypicongpu.json, convert json to dict and extract equation for density
+    # a pypicongpu.json is created for every PICMI call
+    with open("pypicongpu.json") as file:
+        sim_dict = json.load(file)
+        density_fct_str = sim_dict["species_initmanager"]["operations"]["simple_density"][0]["profile"]["data"][
+            "function_body"
+        ]
```

Comment:
Heads-up: Nitpicking coming your way!

I'd probably even unindent that last line. That way the file will only be kept open for as long as it is really necessary. In general, that's a good thing to aim for in order to avoid various risks of file corruption, blocking other code's access, etc. but also to more clearly communicate the dependencies in your code. (`json.load(...)`'s return value might still hold a reference to the file and its `[]` operators could potentially rely on that file still being open.)

In practice, the dictionary look-up afterwards is certainly completely irrelevant in that respect but one might choose following such style rules in the harmless and simple cases, so they come naturally in the dangerous and difficult-to-decipher ones as well.

```suggestion
    with open("pypicongpu.json") as file:
        sim_dict = json.load(file)
    density_fct_str = sim_dict["species_initmanager"]["operations"]["simple_density"][0]["profile"]["data"][
            "function_body"
        ]
```